pageserver/src/tenant/secondary/downloader.rs

```diff
@@ -0,0 +1,791 @@
+                }
+            }
+
+            result_tx
+                .send(CompleteDownload {
+                    secondary_state,
+                    completed_at: Instant::now(),
+                })
+                .ok();
+        }.instrument(info_span!("secondary_download", tenant_id=%tenant_shard_id.tenant_id, shard_id=%tenant_shard_id.shard_slug())));
+        RunningDownload { barrier }
+    }
+}
+
+pub(super) async fn downloader_task(
+    tenant_manager: Arc<TenantManager>,
+    remote_storage: GenericRemoteStorage,
+    command_queue: tokio::sync::mpsc::Receiver<CommandRequest<DownloadCommand>>,
+    background_jobs_can_start: Barrier,
+    cancel: CancellationToken,
+) {
+    // TODO: separate config for downloads
+    let concurrency = tenant_manager.get_conf().heatmap_upload_concurrency;
+
+    let generator = SecondaryDownloader {
+        tenant_manager,
+        remote_storage,
+    };
+    let mut scheduler = Scheduler::new(generator, concurrency);
+
+    scheduler
+        .run(command_queue, background_jobs_can_start, cancel)
+        .instrument(info_span!("secondary_downloads"))
+        .await
+}
+
+/// Scan local storage and build up Layer objects based on the metadata in a HeatMapTimeline
+async fn init_timeline_state(
+    conf: &'static PageServerConf,
+    tenant_shard_id: &TenantShardId,
+    heatmap: &HeatMapTimeline,
+) -> SecondaryDetailTimeline {
+    let timeline_path = conf.timeline_path(tenant_shard_id, &heatmap.timeline_id);
+    let mut detail = SecondaryDetailTimeline::default();
+
+    let mut dir = match tokio::fs::read_dir(&timeline_path).await {
+        Ok(d) => d,
+        Err(e) => {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                let context = format!("Creating timeline directory {timeline_path}");
+                tracing::info!("{}", context);
+                tokio::fs::create_dir_all(&timeline_path)
+                    .await
+                    .fatal_err(&context);
+
+                // No entries to report: drop out.
+                return detail;
+            } else {
+                on_fatal_io_error(&e, &format!("Reading timeline dir {timeline_path}"));
+            }
+        }
+    };
+
+    let heatmap_metadata: HashMap<_, _> = heatmap.layers.iter().map(|l| (&l.name, l)).collect();
+
+    while let Some(dentry) = dir
+        .next_entry()
+        .await
+        .fatal_err(&format!("Listing {timeline_path}"))
+    {
+        let dentry_file_name = dentry.file_name();
+        let file_name = dentry_file_name.to_string_lossy();
+        let local_meta = dentry.metadata().await.fatal_err(&format!(
+            "Read metadata on {}",
+            dentry.path().to_string_lossy()
+        ));
+
+        // Secondary mode doesn't use local metadata files, but they might have been left behind by an attached tenant.
+        if file_name == METADATA_FILE_NAME {
+            continue;
+        }
+
+        match LayerFileName::from_str(&file_name) {
```

Comment:
This code feels like a duplication of the timeline load code.
It indeed needs to be because, to my understanding, the downloader produces a `timelines/` dir that can be picked up by the primary-mode timeline loading function.

I think it's time for an abstraction `TimelineDir` that owns the timeline dir layout and is used by both the secondary downloader and the primary-mode downloader.
Not in this PR, but, should be a follow-up soon after.

Reply:
Yeah, I think to make this happen would be a bit of a refactor of the existing code as well as downloader.  Currently `Layer` and therefore layermap are rather coupled to the attached tenant types (via the `Timeline` reference).

Not sure if it will make sense to try and totally unify them, but it might make sense to define an intermediate lightweight representation (i.e. your `TimelineDir`) that just represents the tree of files, such that we can go `LayerMap`->`TimelineDir`->downloader state

Reply:
> Currently Layer and therefore layermap are rather coupled to the attached tenant types (via the Timeline reference).

Hm, hadn't realized that problem when asking you to carry over the `struct Layer`s between secondary and primary mode.
But, I think it's still the right thing to do: `struct Layer` represents the layer file. It shouldn't have a reference to `Timeline` in the first place. I think it only needs the reference for the index part upload scheduling on drop, which, IIRC, you and I agreed a while back that it shouldn't be part of `struct Layer`.